src/main/kotlin/org/rust/lang/core/macros/MacroExpansionTask.kt

```diff
@@ -5,578 +5,168 @@
-                batch.deleteFile(oldExpansionResult.ok)
-            }
-            return Stage3(info, callHash, defHash, Err(error))
+        val path = "/$MACRO_EXPANSION_VFS_ROOT/$projectDirectoryName"
+        val root = expansionFileSystem.getDirectory(path) ?: return false
+        root.copyChildren()
+            .filterIsInstance<FSItem.FSFile>()
+            .filter { it.name !in crates }
+            .ifEmpty { return false }
+            .forEach { it.delete() }
+
+        expansionFileSystem.findFileByPath(path)?.let {
+            VfsUtil.markDirty(false, false, it)
         }
+        return true
     }
 
-    class Stage3(
-        private val info: ExpandedMacroInfo,
-        private val callHash: HashCode?,
-        private val defHash: HashCode?,
-        private val result: RsResult<Stage3Ok<MacroExpansionVfsBatch.Path>, ExpansionPipelineError>
-    ) : Pipeline.Stage3SaveToStorage {
-        override fun save(storage: ExpandedMacroStorage) {
-            checkWriteAccessAllowed()
-            val result2 = result.map {
-                val file = it.expansionFile.toVirtualFile()
-                    ?: error("(${it.expansionFile}).toVirtualFile() must not be null")
-                // Optimization: skip charset guessing in `VirtualFileImpl.contentToByteArray()`
-                file.setCharset(Charsets.UTF_8, null, false)
-                Stage3Ok(file, it.ranges, it.expansionBytesHash, it.mixHash)
-            }
-            storage.addExpandedMacro(info, callHash, defHash, result2)
-            // If a document exists for expansion file (e.g. when AST tree is loaded), the changes in
-            // a virtual file will not be committed to the PSI immediately. We have to commit it manually
-            // to see the changes (or somehow wait for DocumentCommitThread, but it isn't possible for now)
-            if (result2 is Ok) {
-                val doc = FileDocumentManager.getInstance().getCachedDocument(result2.ok.expansionFile)
-                if (doc != null) {
-                    PsiDocumentManager.getInstance(storage.project).commitDocument(doc)
-                }
-            }
-        }
+    // "<mixHash>_<order>.rs" → "<mixHash>"
+    private fun extractMixHashFromExpansionName(name: String): String {
+        testAssert { name.endsWith(".rs") && name.contains('_') }
+        val index = name.indexOf('_')
+        if (index == -1) return name
+        return name.substring(0, index)
     }
-}
 
-data class Stage3Ok<T>(
-    val expansionFile: T,
-    val ranges: RangeMap,
-    val expansionBytesHash: Long,
-    val mixHash: HashCode,
-)
+    override val waitForSmartMode: Boolean
+        get() = true
 
-private class CorruptedExpansionStorageException : RuntimeException {
-    constructor() : super()
-    constructor(cause: Exception) : super(cause)
+    override val progressBarShowDelay: Int
+        get() = when (taskType) {
+            RsTask.TaskType.MACROS_UNPROCESSED -> 2000
```

Comment:
Previously we did the delay in the opposite cases - `MACROS_FULL` and `MACROS_WORKSPACE`